### PR TITLE
executor: make set transaction as of timestamp supports select statement

### PR DESCRIPTION
**Pull request by Yisaer**
Signed-off-by: yisaer <disxiaofei@163.com>

### What problem does this PR solve?

`set transaction read only as of timestamp` only supports next `begin` transaction.

Problem Summary:

### What is changed and how it works?

This request make it support `select` statement

### Check List 

Tests 

- Unit test

### Release note 

* No release note

### CODE REVIEW

File: planner/core/preprocess.go [hunk 1]
```diff
@@ -1377,6 +1381,11 @@ func (p *preprocessor) checkFuncCastExpr(node *ast.FuncCastExpr) {
 // handleAsOf tries to validate the timestamp.
 // If it is not nil, timestamp is used to get the history infoschema from the infocache.
 func (p *preprocessor) handleAsOf(node *ast.AsOfClause) {
+	readTS := p.ctx.GetSessionVars().TxnReadTS.PeakTxnReadTS()
+	if readTS > 0 && node != nil {
+		p.err = ErrAsOf.FastGenWithCause("can't use select as of while already set transaction as of")
+		return
```
Comment (JmPotato) [inline planner/core/preprocess.go:1387]: So an `AS OF` statement after `SET TRANSACTION READ ONLY AS OF` won't consume the `TxnReadTS` even if it causes an error? Is this expected?

File: executor/simple.go [hunk 2]
```diff
@@ -575,31 +575,21 @@ func (e *SimpleExec) executeBegin(ctx context.Context, s *ast.BeginStmt) error {
 		}
 		if s.AsOf != nil {
 			// start transaction read only as of failed due to we set tx_read_ts before
-			if e.ctx.GetSessionVars().TxnReadTS > 0 {
+			if e.ctx.GetSessionVars().TxnReadTS.PeakTxnReadTS() > 0 {
```
Comment (JmPotato) [inline executor/simple.go:578]: I think using the `used` flag to check rather than `readTS` is safer.

**Review by JmPotato — COMMENTED** [review]
The rest LGTM.

File: planner/core/preprocess.go (hunk 1 above)
Comment (Yisaer) [reply planner/core/preprocess.go:1387]: it is expected.

File: executor/stale_txn_test.go [hunk 3]
```diff
@@ -628,3 +682,35 @@ func (s *testStaleTxnSuite) TestSpecialSQLInStalenessTxn(c *C) {
...
+
+	// confirm schema changed
+	schemaVer2 := tk.Se.GetInfoSchema().SchemaMetaVersion()
+	c.Assert(schemaVer1, Less, schemaVer2)
+	tk.MustExec(fmt.Sprintf(`SET TRANSACTION READ ONLY AS OF TIMESTAMP '%s'`, time1.Format("2006-1-2 15:04:05.000")))
+	c.Assert(tk.Se.GetInfoSchema().SchemaMetaVersion(), Equals, schemaVer1)
+	tk.MustExec("select * from t;")
+	tk.MustExec(fmt.Sprintf(`SET TRANSACTION READ ONLY AS OF TIMESTAMP '%s'`, time1.Format("2006-1-2 15:04:05.000")))
+	tk.MustExec("begin;")
+	c.Assert(tk.Se.GetSessionVars().TxnCtx.InfoSchema.(infoschema.InfoSchema).SchemaMetaVersion(), Equals, schemaVer1)
```
Comment (xhebox) [inline executor/stale_txn_test.go:713]: No, it should still be `tk.Se.GetInfoSchema()`. That is how other code get the infoschema, so do your test program.

Comment (Yisaer) [reply executor/stale_txn_test.go]: updated.

File: executor/simple.go (hunk 2 above)
Comment (Yisaer) [reply executor/simple.go:578]: I think `peak` is more proper here. It is used for the following case:

```sql
set transaction read only as of timestamp ts1
start transaction read only as of timestamp ts2 (expecting error)
begin ( This should be used as ts1 timestamp)
```

**Review by xhebox — APPROVED** [review]

File: planner/core/preprocess.go [hunk 4]
```diff
@@ -101,6 +101,10 @@ func Preprocess(ctx sessionctx.Context, node ast.Node, preprocessOpt ...Preproce
 		v.PreprocessorReturn = &PreprocessorReturn{}
 	}
 	node.Accept(&v)
+	readTS := ctx.GetSessionVars().TxnReadTS.UseTxnReadTS()
+	if readTS > 0 {
```
Comment (nolouch) [inline planner/core/preprocess.go:105]: Will this cause the information scheme not to match with the readTS?

Comment (Yisaer) [reply planner/core/preprocess.go:105]: After we execute `set transaction read only as of timestamp ...`, the infoschema will be stored in `vars.SnapshotInfoschema`

Comment (Yisaer) [reply planner/core/preprocess.go:105]: Also we have a test `TestSetTransactionInfoSchema` to ensure infoschema

Comment (nolouch) [reply planner/core/preprocess.go:105]: There are many places where inforschema is used.  In there, `PreprocessorReturn` will bring the schema, and the set will also `vars.SnapshotInfoschema` store the schema.  maybe we need to sort it in the future, and it's better to have a unified place.

**Review by nolouch — COMMENTED** [review]
/lgtm

**Review by nolouch — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
LGTM

**Pull request merged**